src/3DFactory.R

```diff
@@ -0,0 +1,138 @@
+# LT 17/05/2022
+
+# prototype: 3D Factory AND Adjudicator are in the same source file to avoid
+# implementation of the interface between them
+
+# 3D Factory component
+
+# the 3D Factory generates a list of 3D regions and calculates their intolerance score
+
+# First prototype
+# creates 3D regions of variable length using upper bound of CI
+# each region is built by growing a sphere centered on the carbon-alpha of
+# a specific residue until the whole protein is included
+# a criteria (based on the upper bound) is used to decide the length of the region
+
+# file names and parameters are hard-coded for this prototype
+
+source("constraint_utils.R")
+
+## calculate upper bound of the o/e confidence interval
+get_oe_ci_up <- function(observed, expected, alpha = 0.05) {
+  return(qchisq((1 - alpha / 2), 2 * (observed + 1)) / 2 / expected)
+}
+
+## calculate intolerance metric of a region
+get_region_metric <- function(i_region, alpha, regions, oe_data) {
+  # indices of the residues belonging to this region
+  this_region <- regions[seq_len(i_region)]
+
+  # total number of missense variants observed in the region
+  total_obs <- sum(oe_data$obs[this_region])
+
+  # total number of missense variants expected in the region
+  total_exp <- sum(oe_data$exp[this_region])
+
+  # upper bound of o/e
+  oe_upper <- get_oe_ci_up(total_obs, total_exp, alpha)
+
+  c(obs = total_obs, exp = total_exp, oe = total_obs / total_exp, oe_upper = oe_upper)
+}
+
+### main
+## inputs
+# protein structure file
+prot_file <- "../structures/AF-P68133-F1-model_v2.pdb"
+
+# o/e file, out of the GenomicSimulator or a genomic database
+oe_file <- "../structures/AF-P68133-F1-model_v2.simmut.tsv"
+
+oe <- read.table(
+  file = oe_file, sep = "\t",
+  col.names = c("residue_index", "obs", "exp")
+)
+
+# calculate pairwise distances between residues based on their carbon alpha
+dist_mat <- get_dist_mat(prot_file)
+
+# number of residues in protein
+n <- nrow(dist_mat)
+
+##################
+# 3D Factory
+
+# get 3D region and its metrics for each residue
+regions_list <- lapply(seq_len(n), function(i_residue) {
+
+  # get the corresponding row of the distance matrix
+  row <- dist_mat[i_residue, ]
+
+  # grow the sphere = order by increasing distance
+  this_aa_regions <- seq_len(n)[order(row)]
+
+  # calculate the upper bound of the o/e confidence interval for each region
+  regions_metric <- sapply(seq_len(n), function(i) {
+    get_region_metric(i, 0.2, this_aa_regions, oe_data = oe)

```

Comment:
The easy and not satisfying answer to this question is that alpha = 0.2 is not far from the value used by Konrad Karczewski for LOEUF (alpha = 0.9)
In truth It is largely arbitrary. What alpha controls is the [trade-off between bias and variance](https://towardsdatascience.com/understanding-the-bias-variance-tradeoff-165e6942b229). Low values of alpha will favour bias over variance (i.e it will generate larger regions in size, with the risk of underfitting), larger values of alpha will favour variance over bias, with the risk of overfitting. Here I picked 0.2 because it gave a decent result for the demo.
In the future my plan is to use a proper model selection procedure to choose the 'best' value for alpha. Most likely I will estimate alpha using a maximum likelihood , but not completely sure yet because the statistical model needs to be refined: a Poisson model for the observed value  would underestimate the variability by ignoring the uncertainty on the expected